sos/plugins/juju_controller.py

```diff
@@ -0,0 +1,162 @@
+class JujuController(Plugin, UbuntuPlugin):
+    """ Juju orchestration tool - Controller
+    """
+
+    plugin_name = 'juju_controller'
+    profiles = ('virt', 'sysmgmt')
+
+    # Use files instead of packages here because there is no package installed
+    #  on a juju controller that uniquely identifies it:
+    files = '/var/lib/juju/db'
+
+    option_list = [
+        ('export-mongodb',
+         'Export mongodb collections as json files', '', False,)
+    ]
+
+    def export_mongodb(self, juju_machine_id, dbpass):
+        # We are gonna grab the output of the mongo command
+        # and do some regexp and replace()/split() it to a list.
+        # Usually when connecting with ssl on localhost, the cli send some
+        # output which unfortunately ends up in stdin instead of stderr
+        # and we need to do this extra work to parse it.
+        collections_query = ("mongo 127.0.0.1:37017/juju "
+                             + "--authenticationDatabase admin "
+                             + "--ssl --sslAllowInvalidCertificates --quiet "
+                             + "--username {} --password {} ".format(
+                               juju_machine_id, dbpass)
+                             + "--eval 'db.getCollectionNames();'")
+
+        _collections = self.get_command_output(collections_query)
+
+        if _collections['status'] == 0:
+            collections_output = _collections['output']
+        else:
+            collections_output = ''
+
+        collections_regexp = re.compile(r'(?ms)^\[(.*)\]')
+        collections = collections_regexp.findall(collections_output)[0]
+        subs = [
+            ('\"', ''),
+            (',', ''),
+        ]
+        # In order to be able to properly query the mongodb collection
+        # tables we need to cleanup the output from the collections_query
+        # as those return the " and , symbols, which we cannot cleanly use
+        # those to cleanly query after
+        for search, replace in subs:
+            collections = collections.replace(search, replace)
+        collections = collections.split()
+
+        # We don't collect txns.log, statuseshistory, metrics, txns due to size
+        collections_ignore = ['txns.log', 'statuseshistory', 'metrics', 'txns']
+        for item in collections_ignore:
+            if item in collections:
+                collections.remove(item)
+
+        for collection in collections:
+            filename = "mongoexport_collection_{}.json".format(collection)
+            # This will execute a mongoexport command, note it will add the
+            # password to the sos log, however we remove it from the filename
+            mongoexport_cmd = (
+                "mongoexport --host 127.0.0.1 --port 37017 --db juju "
+                + "--authenticationDatabase admin --ssl "
+                + "--sslAllowInvalidCertificates --username '{}' --password "
+                + "'{}' --collection {} --jsonArray"
```

Comment:
We can't pass passwords like this - as well as appearing in the logs they will be visible to other users on the system in process listings.

For `mongoexport` it looks like the password can be [provided on stdin](https://stackoverflow.com/questions/32152666/automated-mongodump-how-to-hide-password-from-ps-output) - there isn't currently a way for plugins to pass input to external processes but if this is the only option for mongo then we may be able to make this available. An environment variable would be much better however if this is supported.